content/0007-development-environment/rfc.md

```diff
@@ -0,0 +1,231 @@
+Relay SaaS on a developer's computer.
+
+## Motivation
+
+Having a quick and simple tool for bootstrapping, maintaining and updating a
+local cluster will reduce churn in the development cycle and help eliminate the
+deployment of bugs. The testing harness for running workflows kicked off from
+the API is very complicated. This project hopes to reduce the amount of times a
+developer writes some code, commits, waits for CI builds, waits for CD to deploy
+against a remote cluster, then is able to run the code against a real
+environment.
+
+## Product-level explanation
+
+[Explain this change in a way that our product team can understand it and pitch
+it to customers for feedback. Make heavy use of examples and ensure you
+introduce any new concepts at a high level.]
+
+### V1
+
+We introduce a tool called RDE (relay development environment) into the
+deployment repository.
+
+#### RDE
+
+RDE (from now on referred to as `rde`) is a shell script that wraps
+[K3d](https://github.com/rancher/k3d) and is opinionated in how it builds the
+development environment. We avoid the common pit-fall of over-engineering and
+reinventing the wheel by only providing configuration for the most necessary
+things.
+
+If `rde` exists in `$PATH`, then relay repositories use it to update their
+respective deployments with a local build _without_ configuration at all.
+
+Clusters are singleton. There either is or isn't a relay development cluster on
+a machine.
+
+`rde` works on MacOS and Linux as long as a running Docker service is present.
+
+#### Creating/Starting a cluster
+
+Starting and creating, for all intents and purposes, are the same thing. With
+`rde` you invoke the `cluster start` command to ensure a cluster is up and running,
+whether it exists or not.
+
+#### Stopping a cluster
+
+To stop a cluster, you invoke the `cluster stop` command and all the supporting
+containers will be shutdown.
+
+#### Destroying a cluster
+
+To destroy a cluster, you invoke the `cluster delete` command, which will stop
+and destroy running containers and delete any saved metadata and data
+files/directories. This can be used to start a new cluster from scratch.
+
+#### Using `kubectl`
+
+To ensure there's no confusion as to which cluster operations team members are
+accessing to perform maintenance, `kubectl` commands against the development
+cluster are invoked using `rde kubectl`, which knows how to configure kubectl
+with the proper kubeconfig file.
+
+Obviously someone can choose to populate their global kubeconfig with the
+cluster credentials and change context if they want, but any mistakes caused
+doing so is on them.
+
+#### Bootstrapping
+
+* k3d creates a 3 node Kubernetes cluster and `rde` ensures it's running and
+  healthy
+* A kubectl config file is saved the `rde` data directory.
+* Critical 3rd party resources are deployed to the cluster using `kubectl apply`.
+    * tekton
+    * knative
```

Comment:
If we want `kubectl top` and pod autoscaling we'll also need https://github.com/kubernetes-sigs/metrics-server I think?

Reply:
This is actually deployed with k3s